Unity/NativeObject.cs

```diff
@@ -0,0 +1,79 @@
+﻿using System;
+using System.Collections.Specialized;
+
+namespace Unity
+{
+    public unsafe class NativeObject : IDisposable
+    {
+        NativeObjectV1* nativeObject;
+
+        NativeObjectFactory factory;
+
+        PersistentTypeID persistentTypeID;
+
+        public IntPtr* VirtualFunctionTable => nativeObject->VirtualFunctionTable;
+
+        public int InstanceID => nativeObject->InstanceID;
+
+        public IntPtr Pointer => new IntPtr(nativeObject);
+
+        static readonly BitVector32.Section MemLabelIdentifierSection = BitVector32.CreateSection(1 << 11);
+
+        static readonly BitVector32.Section TemporaryFlagsSection = BitVector32.CreateSection(1 << 0, MemLabelIdentifierSection);
+
+        static readonly BitVector32.Section HideFlagsSection = BitVector32.CreateSection(1 << 6, TemporaryFlagsSection);
+
+        static readonly BitVector32.Section IsPersistentSection = BitVector32.CreateSection(1 << 0, HideFlagsSection);
+
+        static readonly BitVector32.Section CachedTypeIndexSection = BitVector32.CreateSection(1 << 10, IsPersistentSection);
+
+
+        public NativeObject(IntPtr ptr, NativeObjectFactory factory, PersistentTypeID persistentTypeID)
+        {
+            if (ptr.ToInt64() == 0) throw new ArgumentException("Object Ptr cannot be null");
+            nativeObject = (NativeObjectV1*)ptr;
+            this.factory = factory;
+            this.persistentTypeID = persistentTypeID;
+        }
+
+        public void Dispose()
+        {
+            if ((IntPtr)nativeObject != IntPtr.Zero)
```

Comment:
Nit: Since this is a raw pointer, it can just be compared to `null` directly: `if (nativeObject != null)`.